MM-61130: Use a channelMember map at web_hub level

agnivade opened the pull request:
Tests at very high scale indicates that the iteration
of all connections during websocket broadcast starts
to become a bottleneck.

To optimize this, we move the channelMember cache from
inside web_conn.go to the hubConnectionIndex.

This involves adding a new map keyed by the channelID
and containing all webConns where the user is a member
of that channel. Subsequently, a new method needed to
be added to invalidate the cache which previously
used to happen in web_conn.

And as a last step, we remove the cache from web_conn
to reduce SQL queries to the DB.

https://mattermost.atlassian.net/browse/MM-61130

Attached are the cpu profiles: 
[cpu_profiles.zip](https://github.com/user-attachments/files/17427714/cpu_profiles.zip)

cpu_150 is old one
cpu_180 is with the changes

Grafana:
OLD:
Threshold: https://snapshots.raintank.io/dashboard/snapshot/mZajUP5Ereb4H55Wmp792BOsMdTMUjQC?orgId=0
V2: https://snapshots.raintank.io/dashboard/snapshot/YiNXaWhhSJGYHjRBbBII1tmlCbWn9fEL?orgId=0

NEW:
Threshold: https://snapshots.raintank.io/dashboard/snapshot/DfqeF49pVt1raQ4jTWzxMmgGXWQjgUbL
load test: (attached)
![sc3](https://github.com/user-attachments/assets/d3dc6b5e-44f1-43cd-879d-381f9d47d78e)
![sc2](https://github.com/user-attachments/assets/196ba756-f001-4024-9a49-d87b47a8b53e)
![sc1](https://github.com/user-attachments/assets/f5a23936-0206-42e2-aec6-3c851f9d1f3d)
![sc6](https://github.com/user-attachments/assets/6bb0fbdd-a2d6-4f1f-9a11-40286c0ee91d)
![sc4](https://github.com/user-attachments/assets/0254d4fd-3b4f-499e-83f9-18b0083c583c)



```release-note
NONE
```

cpoile commented on server/channels/app/platform/web_conn.go:
```diff
@@ -326,6 +323,9 @@ func (wc *WebConn) SetConnectionID(id string) {
 
 // GetConnectionID returns the connection id of the connection.
 func (wc *WebConn) GetConnectionID() string {
+	if wc.connectionID.Load() == nil {
```
A little bit unrelated, but do we _really_ need this to be an atomic? It's only set on connect and reconnect, so maybe we can save some sync costs? (Do we have tests that would detect an error on reconnect if we did remove this?) Undestandable if we don't want to introduce the chance of a race here, but I'm just thinking of ways to squeeze more perf out of this.

agnivade replied:
Hmm .. I need to check this. Atomics are usually extremely fast, and this has never appeared in the profiles. But I am totally supportive of not using atomics unless we need to. I'll look into this separately.

cpoile reviewed (approved):
This is some nice code, thanks Agniva!
The only observation I would have is that I'm worried the tests are really stressing the combinations of maps we now have. (Unless that's somewhere else?) It would be nice to see, say, a function that took 100 users over 10 channels and added and removed them hundreds of times randomly, and then asserted that the final combination of users/channels is correctly stored in the various maps. Or maybe that's too idealistic. Anyway, not a req. Nice work!

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -619,30 +630,55 @@ func areAllInactive(conns []*WebConn) bool {
+		// In this case, the map won't be populated. This is effectively the same
+		// as not broadcasting the event, and that's how it was implemented in
+		// platform/web_conn.go
+		i.logger.Warn("Error while getting GetAllChannelMembersForUser", mlog.Err(err))
```
I am still not happy with this. Thinking more, if this failed inside web_conn.go, only one connection gets affected, but if this fails here, then a large portion of users will be affected. We cannot continue with a failure. Will improve this.

agnivade commented:
>  It would be nice to see, say, a function that took 100 users over 10 channels and added and removed them hundreds of times randomly, and then asserted that the final combination of users/channels is correctly stored in the various maps. Or maybe that's too idealistic. 

This is a good idea. The problem is app/platform doesn't have (or rather, difficult to add) the necessary helper functions (due to refactors) to do such high level things. I even needed to copy-paste the store method to add a channel member, because suiteIface didn't have that.

I will add a test from api layer to check this. Good point.

agnivade commented:
Ok, this is ready to review now, sans the high level additional tests with cpoile asked for.

streamer45 commented:
@agnivade A meta request for this would be (if possible) to link to some dashboards (i.e., Grafana snapshot) or supporting data from scaling tests you are conducting (e.g., pprof profiles).

agnivade commented:
@streamer45 - Done. Attached the results in the PR message.

agnivade commented:
Added high level test @cpoile. PTAL.

cpoile reviewed (approved):
I'm assuming there's no chance that adding and removing multiple users would cause inconsistency, right? I think this is good then, thanks Agniva!

agnivade commented:
It's really the same piece of code that gets executed. Don't think that increases code coverage anyway. Important thing is that we covered the add and remove paths.

agarciamontoro commented on server/channels/app/platform/web_hub.go:
```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
+		i.byChannelID[chID] = filtered
+	}
+
+	// re-populate the cache
+	cm, err := i.getChannelMembersForUser(userID)
+	if err != nil {
+		panic(fmt.Errorf("error while getting GetAllChannelMembersForUser: %v", err))
+	}
```
We're `panic`-ing here as well, I think that's another argument to `panic` inside the function.

agnivade replied:
Yep, good point.

agarciamontoro commented on server/channels/app/platform/web_hub.go:
```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
+		if err == nil {
+			return cm, nil
+		}
+		// It is imperative that we don't block the hub
+		time.Sleep(time.Millisecond)
```
Are we concerned about this millisecond here? Also, should we log the first error in the query in case it does affect performance?

agnivade replied:
Not particularly concerned. It's a hit we need to take. Will log the first error. Thanks.

streamer45 replied:
Not sure I fully understand the use of `Sleep` here. The caller of `getChannelMembersForUser` is blocked regardless, so what exactly are we unblocking by sleeping?

agnivade replied:
By letting the DB some time to recover from whatever it is failing from.

streamer45 replied:
I see. I find the comment a bit unclear if the purpose of the sleep is to let the DB "breath". Plus I am not sure one millisecond would help that much since we are talking about a couple of retries so if it failed a ms before, I say there's a good chance it could fail again. And now I wonder, how does the retry store layer fit into this? In that case we pause for 100ms which sounds more reasonable.

agnivade replied:
The retry store layer is purely for retriable queries. e.g: ones which got cancelled due to a deadlock. Those are safe to be auto-retried. This is more about if the DB failed to reply due to load or something. So what you are suggesting is to remove the retry completely? I'd be fine with that.

streamer45 replied:
Yes, it could be simpler if we don't retry at all and fail fast. Otherwise, we'd have to bump the sleep to a more meaningful value, I feel, and block for longer as a result.

agnivade replied:
Done.

agarciamontoro commented on server/channels/app/platform/web_hub.go:
```diff
@@ -693,6 +790,20 @@ func (i *hubConnectionIndex) ForUser(id string) []*WebConn {
 
+// ForChannel returns all connections for a channelID.
+func (i *hubConnectionIndex) ForChannel(channelID string) []*WebConn {
+	if len(i.byChannelID[channelID]) <= 1 {
+		return i.byChannelID[channelID]
+	}
+
+	// Note: this is expensive because usually there will be
+	// more than 1 member for a channel, and broadcasting
+	// is a hot path, but worth it.
+	conns := make([]*WebConn, len(i.byChannelID[channelID]))
+	copy(conns, i.byChannelID[channelID])
+	return conns
```
Why do we have a special case for `len(i.byChannelID[channelID]) <= 1`? If we need to clone it, why not clone it there as well?

agnivade replied:
Because the slice length is not re-adjusted there. Since it has just one element. Refer to the PR: https://github.com/mattermost/mattermost/pull/25785

agarciamontoro reviewed (approved):
Just a couple of questions and nits below. This is a great improvement!

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
+
+const numRetries = 2
+
+// getChannelMembersForUser is a wrapper around the store method which retries
+// for numRetries times. After that we panic, to fail loudly because it is even
+// worse to continue without the channel membership data.
+// Note: the hub has internal recovery mechanism to auto-restart in case of a panic.
+// So only one hub out of numCPU hubs will be affected.
```
I'd prefer if we prevented the specific user from connecting rather than crashing the hub along with all connections (from multiple users) in them.

agnivade replied:
That will be difficult at the hub level because this data is used to control broadcast data for all users in the hub. The per-user control was possible when this was at the web_conn level.

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
+
+func (i *hubConnectionIndex) InvalidateCMCacheForUser(userID string) {
+	// Clear out all user entries which belong to channels.
+	for chID, webConns := range i.byChannelID {
```
This code seems practically identical to the above, except for the filtering condition. Wondering if it'd be valuable to refactor into a single util.

agnivade replied:
Let's do it if this gets repeated one more time. Following the rule of 3.

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
+	}
+}
+
+const numRetries = 2
```
Maybe I'd move this on top of the file and give it a slightly more verbose name to clearly indicate the context of its purpose.

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -693,6 +790,20 @@ func (i *hubConnectionIndex) ForUser(id string) []*WebConn {
 
+// ForChannel returns all connections for a channelID.
+func (i *hubConnectionIndex) ForChannel(channelID string) []*WebConn {
+	if len(i.byChannelID[channelID]) <= 1 {
+		return i.byChannelID[channelID]
+	}
```
Wondering if the fast path here is adding much value. It feels like additional code to cover an unrealistic case that would be fast regardless. Thoughts?

agnivade replied:
Makes sense. Removing.

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -497,6 +502,7 @@ func (h *Hub) Start() {
 				for _, webConn := range connIndex.ForUser(userID) {
 					webConn.InvalidateCache()
 				}
+				connIndex.InvalidateCMCacheForUser(userID)
```
To confirm, this is the only place where we need to invalidate, and it would trigger if a user joined a new channel or left an existing one (and a few more cases).

agnivade replied:
Correct. We are just doing it at the same level as the `webConn.InvalidateCache`, because we are basically hoisting up the membership check from webConn level to webHub level.

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -620,30 +632,51 @@ func areAllInactive(conns []*WebConn) bool {
+		panic(fmt.Errorf("error while getting GetAllChannelMembersForUser: %v", err))
+	}
+	for chID := range cm {
+		i.byChannelID[chID] = append(i.byChannelID[chID], wc)
```
Mostly a curiosity of mine: do we have heap profiles around to understand how heavy this map can get?

agnivade replied:
Very minimal. 4.25MB. From the latest test.
[heap_200K.zip](https://github.com/user-attachments/files/17537000/heap_200K.zip)

streamer45 reviewed (commented):
Great improvement. Left a few comments and thoughts.

I'd also recommend security and qa reviews since we are touching the scope of ws events, but will defer to you on that.

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -557,8 +563,14 @@ func (h *Hub) Start() {
 					}
-				} else if msg.GetBroadcast().UserId != "" {
-					candidates := connIndex.ForUser(msg.GetBroadcast().UserId)
+				} else if userID := msg.GetBroadcast().UserId; userID != "" {
+					candidates := connIndex.ForUser(userID)
+					for _, webConn := range candidates {
+						broadcast(webConn)
+					}
+					continue
```
Done

agnivade commented:
@streamer45 - PTAL. This has become more complicated now. I ran into situations where a small mistake would lead to a hub deadlock in trying to force close the connection.

@agarciamontoro - requesting another review as well because the code has changed a lot after Claudio's suggestion to force-close the user connection instead of panicking.

I am still not 100% confident because this code is one of the most complicated areas in MM. But let me know how it looks to you guys. We definitely need a thorough QA and security review on this.

agnivade commented:
Hmm, ok the tests are stuck. There's probably a deadlock somewhere.

agnivade commented:
I cancelled the tests after 50 minutes. Pretty sure there's a deadlock somewhere. 

 @streamer45 @agarciamontoro - What are your thoughts on dropping this PR? Do you think the gain is worth the risk?

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -265,11 +273,17 @@ func (ps *PlatformService) WebConnCountForUser(userID string) int {
+func (h *Hub) Register(webConn *WebConn) error {
+	wr := &webConnRegisterMessage{
+		conn: webConn,
+		err:  make(chan error),
```
It looks safe as is, but how about making this buffered to avoid blocking the hub even for a moment?

agnivade replied:
This feels like a micro-optimization. I want to lean on the side of correctness for now.

streamer45 commented on server/channels/app/platform/web_hub.go:
```diff
@@ -503,7 +522,13 @@ func (h *Hub) Start() {
+				if err != nil {
+					h.platform.Log().Error("Error while invalidating channel member cache", mlog.String("user_id", userID), mlog.Err(err))
+					for _, webConn := range connIndex.ForUser(userID) {
+						webConn.Close()
```
This could be problematic as I think closing would call `(h *Hub) Unregister` before returning.

agnivade replied:
You are right. Following the existing method of :

```
close(webConn.send)
connIndex.Remove(webConn)
```

streamer45 reviewed (commented):
Thanks. I understand that working with mocks has been painful on this one but I also think it would be valuable to have test cases for `GetAllChannelMembersForUser` failing. Anything we can do about that?

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -497,6 +520,14 @@ func (h *Hub) Start() {
+				err := connIndex.InvalidateCMCacheForUser(userID)
+				if err != nil {
+					h.platform.Log().Error("Error while invalidating channel member cache", mlog.String("user_id", userID), mlog.Err(err))
+					for _, webConn := range connIndex.ForUser(userID) {
```
Yes, we copy the slice, so it should be fine.

agnivade commented on server/channels/app/platform/web_hub.go:
```diff
@@ -524,7 +524,8 @@ func (h *Hub) Start() {
 					h.platform.Log().Error("Error while invalidating channel member cache", mlog.String("user_id", userID), mlog.Err(err))
 					for _, webConn := range connIndex.ForUser(userID) {
-						webConn.Close()
+						close(webConn.send)
+						connIndex.Remove(webConn)
```
Fair enough :P

agnivade replied:
Hmm .. I only see this 2 times. Where are you seeing the 3rd time?

agnivade replied:
Ah nvm :facepalm:

agnivade commented:
>  I understand that working with mocks has been painful on this one but I also think it would be valuable to have test cases for GetAllChannelMembersForUser failing. Anything we can do about that?

This would be really challenging. Assigning the mock store to the hub causes all sorts of race conditions which are hard to eradicate. I personally tested this by randomly failing the store method once every 4 calls and verified that the websocket reconnects fine. Did this both for connect and invalidate paths.

agnivade commented:
Ok, added a test, and found a bug! :scream:

agnivade commented on server/channels/api4/post_test.go:
```diff
@@ -3063,6 +3064,30 @@ func TestWebHubMembership(t *testing.T) {
+	_, _, err := cli.Login(context.Background(), th.BasicUser.Username, th.BasicUser.Password)
+	require.NoError(t, err)
+
+	_, err = th.Server.Store().GetInternalMasterDB().Exec(`ALTER TABLE ChannelMembers RENAME to dummy`)
```
Indeed. :D

agnivade commented:
Gentle ping @agarciamontoro.

agarciamontoro reviewed (approved):
Changes look good to me, I'm happy we're no longer `panic`ing :)

agnivade commented:
/e2e-test

DHaussermann commented:
@streamer45 and @agnivade I'm having a setup issue with this. As per test steps provided in the Jira, I need my local HA setup working but it does not want to start the way it has in the past for other PRs
![image](https://github.com/user-attachments/assets/5b67ca51-60ee-4480-86bb-2f341b06cf65)

I'm confused because I think these are different images that run in AMD64 so what is unique about the inbucket image?

0/5 Is there a quick way I can remove inbucket from the setup to see if it resolves the problem or is it required for HA?

DHaussermann commented:
Regarding the setup issue ☝️ the local HA command seems unreliable. I got this working simply by trying the same thing repeatedly. I don't know why it sometimes cause a problem that inbucket doesn't work and sometimes does. 
Also the Redis container sometimes does not start for no apparent reason and I must start it manually and then restart the server.

DHaussermann reviewed (approved):
Tested and passed
- I tested this locally on HA having 3 - 4 unique users and sessions and monitoring WS events
- Completed all actions listed in the jira task for add/remove users to teams and channels
- Using events such as the `typing`, I ensured events were scoped properly based on channel membership and team membership
- Ensured no issues after a connection drop and restore
- Repeated testing with Redis disabled
No issues found. 

LGTM!

agnivade commented:
Thanks Dylan!

> Regarding the setup issue ☝️ the local HA command seems unreliable. 

Could you create a post in hub in PDE private discussion and see what other people say? I don't use the Makefile commands, but if this is unreliable, then we definitely need to make this correct.
